src/js/components/MainNav/MainNav.js

```diff
@@ -1,30 +1,95 @@
 import React, { useContext } from 'react';
-import { withRouter } from 'react-router';
 import '../../../scss/components/MainNav/MainNav.scss';
 import { GlobalContext } from '../../contexts/GlobalContextWrapper';
 import AddServerButton from '../AddServerButton/AddServerButton';
 import MainNavButton from '../MainNavButton/MainNavButton';
 import ServerLink from '../ServerLink/ServerLink';
+import useDiscordData from '../../hooks/useDiscordData';
 
-/**
- * Let's wrap the component with withRouter() to make it receptive to route changes
- */
-const MainNav = withRouter(() => (
-  <div className="nav-column nav-column-server">
-    <div className="nav-group">
-      <MainNavButtons />
+const MainNav = () => {
+  const { Config } = useContext(GlobalContext);
+
+  const config = Config.get([
+    'discordAPIResources',
+    'mainNavButtons',
+    'mainNavButtonTypes',
+    'discordUrls',
+  ]);
+
+  const { discordAPIResources, mainNavButtons, mainNavButtonTypes, discordUrls } = config;
+  const { link, addServerButton } = mainNavButtonTypes;
+  const { client, guilds, icons } = discordAPIResources;
+  const { appCDN } = discordUrls;
+
+  const guildData = useDiscordData(guilds, client);
+
+  let addGuildData;
+
+  if (guildData) {
+    // Sort guilds in alphabetical order
+    guildData.sort((a, b) => {
+      const nameA = a.name.toUpperCase();
+      const nameB = b.name.toUpperCase();
+      if (nameA < nameB) {
+        return -1;
+      }
+      if (nameA > nameB) {
+        return 1;
+      }
+      return 0;
+    });
+
+    // Add sort, button type, and the icon url to each guild
+    addGuildData = guildData.map((data, index) => {
+      const { id, icon } = data;
+
+      const iconUrl = icon ? [appCDN, icons, id, icon].join('/') : null;
+
+      return {
+        ...data,
+        icon: iconUrl,
+        sort: index + 1,
+        type: link,
+      };
+    });
+
+    // Find the home and utility buttons from `mainNavButtons`
+    const homeButton = mainNavButtons.filter((e) => !e.sort);
+    let utilityButtons = mainNavButtons.filter((e) => !e.type || e.type === addServerButton);
+
+    const utilityButtonPosition = addGuildData.length;
+
+    // Changing the position of the utility buttons depending on the number of guilds
+    utilityButtons = utilityButtons.map((e) => ({
+      ...e,
+      sort: e.sort + utilityButtonPosition,
+    }));
+
+    // Set the new buttons in "Config"
+    if (homeButton && addGuildData && utilityButtons) {
+      Config.set({
+        mainNavButtons: [
+          ...homeButton,
+          ...addGuildData,
+          ...utilityButtons,
+        ],
+      });
```

Comment:
So does this still need to go back to config? Or is there a way to do it with cleaner separation of concerns?